### PR TITLE
feat(dunning): Add DunningCampaign resolver to graphql api

### PR DESCRIPTION
**Pull request by ancorcruz**
## Roadmap Task

👉 https://getlago.canny.io/feature-requests/p/set-up-payment-retry-logic

👉 https://getlago.canny.io/feature-requests/p/send-reminders-for-overdue-invoices

 ## Context

We want to automate dunning process so that our users don't have to look at each customer to maximize their chances of being paid retrying payments of overdue balances and sending email reminders.

We are extending dunning campaigns management to edit and delete campaigns.

 ## Description

This change adds a new resolver to graphql api to retrieve a single dunning campaign by id. This is required to render the dunning campaign edit screen in the frontend side.

Also adds counter cache customers_count on dunning campaigns customer relationship to provide a performant count of customers with the dunning campaign applied.

### CODE REVIEW

File: app/graphql/resolvers/dunning_campaign_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+module Resolvers
+  class DunningCampaignResolver < Resolvers::BaseResolver
+    include AuthenticableApiUser
+    include RequiredOrganization
+
+    description "Query a single dunning campaign of an organization"
+
+    REQUIRED_PERMISSION = "dunning_campaigns:view"
+
+    argument :id, ID, required: true, description: 'Unique ID of the integration collection mappings'
```
Comment (rsempe) [inline app/graphql/resolvers/dunning_campaign_resolver.rb:12]:
```suggestion
    argument :id, ID, required: true, description: 'Unique ID of the dunning campaign'
```

File: app/models/customer.rb [hunk 2]
```diff
@@ -192,6 +195,14 @@ def ensure_slug
 
     self.slug = "#{organization.document_number_prefix}-#{formatted_sequential_id}"
   end
+
+  def decrement_customers_count
+    applied_dunning_campaign&.increment!(:customers_count, -1) # rubocop:disable Rails/SkipsModelValidations
```
Comment (rsempe) [inline app/models/customer.rb:200]:
```suggestion
    applied_dunning_campaign&.decrement!(:customers_count) # rubocop:disable Rails/SkipsModelValidations
```

File: spec/graphql/resolvers/dunning_campaign_resolver_spec.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        permissions: required_permission,
+        query:,
+        variables: {dunningCampaignId: dunning_campaign.id}
+      )
+
+      expect_graphql_error(result:, message: "Not in organization")
+    end
+  end
+
+  context "without current organization" do
```
Comment (rsempe) [inline spec/graphql/resolvers/dunning_campaign_resolver_spec.rb:104]: is it useful? we already have `it_behaves_like "requires current organization"`

File: spec/graphql/resolvers/dunning_campaign_resolver_spec.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      )
+
+      expect_graphql_error(
+        result:,
+        message: "Resource not found"
+      )
+    end
+  end
+
+  context "when not member of the organization" do
```
Comment (rsempe) [inline spec/graphql/resolvers/dunning_campaign_resolver_spec.rb:90]: is it useful? we already have it_behaves_like "requires current organization"

File: app/graphql/types/dunning_campaigns/object.rb [hunk 5]
```diff
@@ -9,6 +9,7 @@ class Object < Types::BaseObject
 
       field :applied_to_organization, Boolean, null: false
       field :code, String, null: false
+      field :customers_count, Integer, null: false
```
[the remark this answers is not in the document]
Comment (ancorcruz) [reply app/graphql/types/dunning_campaigns/object.rb:12]: updated, it now includes both, customers with the campaign assigned explicitly plus the customers with the organization's default fallback when the campaign is applied to the organization.

File: spec/graphql/resolvers/dunning_campaign_resolver_spec.rb (hunk 3 above)
Comment (ancorcruz) [reply spec/graphql/resolvers/dunning_campaign_resolver_spec.rb]: you are right, I have dropped the last to scenarios as both are already included in the shared example. Thanks.

**Review by rsempe — APPROVED** [review]
👏

**Pull request merged**